uaclient/apt_news.py

```diff
@@ -0,0 +1,153 @@
+import datetime
+import logging
+import unicodedata
+from typing import List, Optional
+
+from uaclient import util
+from uaclient.config import UAConfig
+from uaclient.data_types import (
+    BoolDataValue,
+    DataObject,
+    DatetimeDataValue,
+    Field,
+    StringDataValue,
+    data_list,
+)
+from uaclient.files import state_files
+
+
+# TODO: list of selectors is undecided
+class AptNewsMessageSelectors(DataObject):
+    fields = [
+        Field("pro", BoolDataValue, required=False),
+    ]
+
+    def __init__(self, pro: bool):
+        self.pro = pro
+
+
+class AptNewsMessage(DataObject):
+    fields = [
+        Field("begin", DatetimeDataValue),
+        Field("end", DatetimeDataValue, required=False),
+        Field("selectors", AptNewsMessageSelectors, required=False),
+        Field("lines", data_list(StringDataValue)),
+    ]
+
+    def __init__(
+        self,
+        begin: datetime.datetime,
+        end: Optional[datetime.datetime],
+        selectors: Optional[AptNewsMessageSelectors],
+        lines: List[str],
+    ):
+        self.begin = begin
+        self.end = end
+        self.selectors = selectors
+        self.lines = lines
+
+
+def is_control_char(c: str) -> bool:
+    return unicodedata.category(c)[0] == "C"
+
+
+def clean_line(line: str) -> str:
+    line_without_control_chars = "".join(
+        [c for c in line if not is_control_char(c)]
```

Comment:
This is just a safe mechanism right ? Because we shouldn't allow control characters on the apt news messages in the first place

Reply:
Right, but we can't trust the server to be strict (the url is configurable) so we do need these kinds of checks in the client as well.